src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/CrossChainTransferStore.cs

```diff
@@ -177,155 +204,234 @@ private HashHeightPair TipToChase()
-                            await this.PutTransferAsync(dbreezeTransaction, partialTransfer);
-                        }
+                        this.PutTransfer(dbreezeTransaction, kv.Key);
                     }
 
                     dbreezeTransaction.Commit();
+
+                    this.UpdateLookups(tracker);
+                }
+                catch (Exception err)
+                {
+                    this.logger.LogError("Error during database update: {0}", err.Message);
+                    this.logger.LogTrace("(-):[SANITY_ERROR]");
+
+                    // Restore expected store state in case the calling code retries / continues using the store.
+                    dbreezeTransaction.Rollback();
+                    throw err;
                 }
             }
+        }
 
-            this.logger.LogTrace("(-)");
+        public static Transaction BuildDeterministicTransaction(uint256 depositId, Recipient.Recipient recipient, Network network,
+            IFederationWalletTransactionHandler federationWalletTransactionHandler, string walletPassword = "", Transaction[] partials = null, ILogger logger = null)
+        {
+            try
+            {
+                logger?.LogTrace("()");
+
+                uint256 opReturnData = depositId;
+
+                // Build the multisig transaction template.
+                var multiSigContext = new Recipient.TransactionBuildContext(new[] { recipient }.ToList(), opReturnData: opReturnData.ToBytes())
+                {
+                    TransactionFee = Money.Coins(0.01m), // TODO: Configurable?
+                    MinConfirmations = 1,                // TODO: This may have to be zero to allow multiple deposits spending a single UTXO.
+                    Shuffle = false,
+                    IgnoreVerify = true,
+                    WalletPassword = walletPassword,
+                    Sign = (walletPassword ?? "") != ""
+                };
+
+                // Build the transaction.
+                Transaction transaction = federationWalletTransactionHandler.BuildTransaction(multiSigContext);
+
+                if (partials?.Any() ?? false)
+                {
+                    multiSigContext.TransactionBuilder.CombineSignatures(partials);
+                }
+
+                logger?.LogTrace("(-)");
+
+                return transaction;
+            }
+            catch (Exception error)
+            {
+                logger?.LogTrace("Could not create transaction for deposit {0}: {1}", depositId, error.Message);
+            }
+
+            return null;
         }
 
         /// <inheritdoc />
-        public async Task RecordLatestMatureDepositsAsync(IDeposit[] deposits)
+        public Task RecordLatestMatureDepositsAsync(IDeposit[] deposits)
         {
             Guard.NotNull(deposits, nameof(deposits));
             Guard.Assert(!deposits.Any(d => d.BlockNumber != this.NextMatureDepositHeight));
```

Comment:
This is fine - however, if the assert is true would it be better to log the fact ?  Question is would this information be important.

Reply:
It should probably be handled the same as any other case where the programmer passes invalid or incomplete information to a method - i.e. with `Guard`. It is not something that should normally occur.

Reply:
Ok, thanks.